newIDE/app/src/UI/TreeView/TreeViewRow.js

```diff
@@ -0,0 +1,458 @@
+        }}
+        onClick={stopPropagation}
+        onDoubleClick={stopPropagation}
+        onBlur={() => {
+          onEndRenaming(value);
+        }}
+        onKeyDown={onInputKeyDown}
+        onKeyUp={e => {
+          if (shouldCloseOrCancel(e)) {
+            // Prevent closing dialog if TreeView is displayed in dialog.
+            e.preventDefault();
+            onEndRenaming(initialValue);
+          } else if (shouldValidate(e)) {
+            onEndRenaming(value);
+          }
+        }}
+      />
+    </div>
+  );
+};
+
+const memoized = memoizeOne((initialValue, getContainerYPosition) =>
+  getContainerYPosition()
+);
+
+type Props<Item> = {|
+  index: number,
+  style: any,
+  data: ItemData<Item>,
+  /** Used by react-window. */
+  isScrolling?: boolean,
+|};
+
+const TreeViewRow = <Item: ItemBaseAttributes>(props: Props<Item>) => {
+  const { data, index, style } = props;
+  const {
+    flattenedData,
+    onOpen,
+    onSelect,
+    onBlurField,
+    onStartRenaming,
+    onEndRenaming,
+    renamedItemId,
+    onContextMenu,
+    canDrop,
+    onDrop,
+    onEditItem,
+    isMobileScreen,
+    DragSourceAndDropTarget,
+    getItemHtmlId,
+  } = data;
+  const node = flattenedData[index];
+  const left = node.depth * 15;
+  const [isStayingOver, setIsStayingOver] = React.useState<boolean>(false);
+  const [
+    hasDelayPassedBeforeEditingName,
+    setHasDelayPassedBeforeEditingName,
+  ] = React.useState<boolean>(false);
+  const openWhenOverTimeoutId = React.useRef<?TimeoutID>(null);
+  const [whereToDrop, setWhereToDrop] = React.useState<
+    'before' | 'after' | 'inside'
+  >('before');
+  const containerRef = React.useRef<?HTMLDivElement>(null);
+  const openContextMenu = React.useCallback(
+    ({ clientX, clientY }) => {
+      onContextMenu({
+        index: index,
+        item: node.item,
+        x: clientX,
+        y: clientY,
+      });
+    },
+    [onContextMenu, index, node.item]
+  );
+
+  const longTouchForContextMenuProps = useLongTouch(openContextMenu, {
+    delay: 1000,
+  });
+
+  const onClick = React.useCallback(
+    event => {
+      if (!node || node.item.isPlaceholder) return;
+      if (node.item.isRoot) {
+        onOpen(node);
+        return;
+      }
+      onSelect({ node, exclusive: !(event.metaKey || event.ctrlKey) });
+    },
+    [onSelect, node, onOpen]
+  );
+
+  const selectAndOpenContextMenu = React.useCallback(
+    (event: MouseEvent) => {
+      onClick(event);
+      openContextMenu(event);
+    },
+    [onClick, openContextMenu]
+  );
+
+  /**
+   * Effect that opens the node if the user is dragging another node and stays
+   * over the node.
+   */
+  React.useEffect(
+    () => {
+      if (
+        isStayingOver &&
+        !openWhenOverTimeoutId.current &&
+        node.canHaveChildren &&
+        node.collapsed
+      ) {
+        openWhenOverTimeoutId.current = setTimeout(() => {
+          onOpen(node);
+        }, 800);
```

Comment:
should those values be extracted at the top of the file? So we can see the different time triggers for actions

Reply:
Good idea